users/tests/test_pipeline.py

```diff
@@ -0,0 +1,94 @@
+import hashlib
+from unittest.mock import Mock
+
+import pytest
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+
+from users.pipeline import associate_by_email, set_default_nickname
+
+User = get_user_model()
+
+
+@pytest.mark.django_db()
+def test_associate_by_email_existing_user() -> None:
+    email = "existing@example.com"
+    nickname = "ExistingNick"
+    existing_user = User.objects.create(email=email, nickname=nickname)
+    details = {"email": email}
+
+    result = associate_by_email(strategy=Mock(), details=details, backend=Mock(), user=None)
+
+    assert result["user"] == existing_user
+    assert result["user"].nickname == nickname
+
+
+@pytest.mark.django_db()
+def test_associate_by_email_no_user_found() -> None:
+    # Prepare the data with a non-existing email
```

Comment:
not needed comment - if really want it, use this in email like not@existing.user.com